From 1c18ba639acd19ba110c49e7eabfbda4d0377c2b Mon Sep 17 00:00:00 2001
From: Valentin Caron <valentin.caron@foss.st.com>
Date: Tue, 18 Jun 2024 15:50:44 +0200
Subject: [PATCH 0965/1695] plat-stm32mp2: add stm32mp21 support in power
 drivers

VDDIO4, UCPD and GPU regulators are not available on STM32MP21 soc.

VCORE and VCPU monitoring are not available on STM32MP21 soc.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 004ca1e29a266efd09a1073d2575d7a0333daa02 plat-stm32mp2: add stm32mp21 support in power drivers]
Signed-off-by: Valentin Caron <valentin.caron@foss.st.com>
Change-Id: I5de6d006a81c1996bbac1c606c7f079336f9b8d0
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/384507
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
---
 .../arm/plat-stm32mp2/drivers/stm32mp25_pwr.c |  3 +++
 .../drivers/stm32mp25_pwr_regul.c             | 20 ++++++++++++++++++-
 2 files changed, 22 insertions(+), 1 deletion(-)

diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.c
index d463263f7..9e5dff584 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr.c
@@ -336,15 +336,18 @@ void stm32mp_pwr_monitoring_enable(enum pwr_monitoring monitoring)
 	vaddr_t pwr_base = stm32_pwr_base();
 
 	switch (monitoring) {
+	/* On MP21 V08CAP is called VBAT */
 	case PWR_MON_V08CAP_TEMP:
 		io_setbits32(pwr_base + _PWR_CR2, _PWR_CR2_MONEN);
 		break;
+#ifndef CFG_STM32MP21
 	case PWR_MON_VCORE:
 		io_setbits32(pwr_base + _PWR_CR5, _PWR_CR5_VCOREMONEN);
 		break;
 	case PWR_MON_VCPU:
 		io_setbits32(pwr_base + _PWR_CR6, _PWR_CR6_VCPUMONEN);
 		break;
+#endif /* !CFG_STM32MP21 */
 	default:
 		break;
 	}
diff --git a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_regul.c b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_regul.c
index a6ef28c56..df9459f18 100644
--- a/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_regul.c
+++ b/core/arch/arm/plat-stm32mp2/drivers/stm32mp25_pwr_regul.c
@@ -425,11 +425,17 @@ enum pwr_regulator {
 	IOD_VDDIO1,
 	IOD_VDDIO2,
 	IOD_VDDIO3,
+#ifndef CFG_STM32MP21
 	IOD_VDDIO4,
+#endif
 	IOD_VDDIO,
+#ifndef CFG_STM32MP21
 	REGU_UCPD,
+#endif
 	REGU_A,
+#ifndef CFG_STM32MP21
 	REGU_GPU,
+#endif
 	PWR_REGU_COUNT
 };
 
@@ -461,6 +467,7 @@ static struct pwr_regu pwr_regulators[PWR_REGU_COUNT] = {
 		.is_an_iod = true,
 		.comp_idx = SYSFG_VDDIO3_ID,
 	 },
+#ifndef CFG_STM32MP21
 	 [IOD_VDDIO4] = {
 		.enable_reg = PWR_CR1,
 		.enable_mask = PWR_CR1_VDDIO4VMEN,
@@ -470,24 +477,28 @@ static struct pwr_regu pwr_regulators[PWR_REGU_COUNT] = {
 		.is_an_iod = true,
 		.comp_idx = SYSFG_VDDIO4_ID,
 	 },
+#endif
 	 [IOD_VDDIO] = {
 		.enable_reg = PWR_CR1,
 		.vrsel_mask = PWR_CR1_VDDIOVRSEL,
 		.is_an_iod = true,
 		.comp_idx = SYSFG_VDD_IO_ID,
 	 },
+#ifndef CFG_STM32MP21
 	 [REGU_UCPD] = {
 		.enable_reg = PWR_CR1,
 		.enable_mask = PWR_CR1_UCPDVMEN,
 		.ready_mask = PWR_CR1_UCPDRDY,
 		.valid_mask = PWR_CR1_UCPDSV,
 	 },
+#endif
 	 [REGU_A] = {
 		.enable_reg = PWR_CR1,
 		.enable_mask = PWR_CR1_AVMEN,
 		.ready_mask = PWR_CR1_ARDY,
 		.valid_mask = PWR_CR1_ASV,
 	 },
+#ifndef CFG_STM32MP21
 	 [REGU_GPU] = {
 		.enable_reg = PWR_CR12,
 		.enable_mask = PWR_CR12_GPUVMEN,
@@ -496,6 +507,7 @@ static struct pwr_regu pwr_regulators[PWR_REGU_COUNT] = {
 		.keep_monitor_on = true,
 		.rifsc_filtering_id = STM32MP25_RIFSC_GPU_ID,
 	 },
+#endif
 };
 
 #define DEFINE_REGUL(_id, _name, _supply) {		\
@@ -518,15 +530,21 @@ static struct pwr_regu pwr_regulators[PWR_REGU_COUNT] = {
 static struct regulator pwr_regu_device[PWR_REGU_COUNT];
 
 /* Not const to allow probe to reassign ops if device is a fixed_iod */
-static const struct regu_dt_desc pwr_regu_desc[] = {
+static const struct regu_dt_desc pwr_regu_desc[PWR_REGU_COUNT] = {
 	[IOD_VDDIO1] = DEFINE_REGUL(IOD_VDDIO1, "vddio1", "vddio1"),
 	[IOD_VDDIO2] = DEFINE_REGUL(IOD_VDDIO2, "vddio2", "vddio2"),
 	[IOD_VDDIO3] = DEFINE_REGUL(IOD_VDDIO3, "vddio3", "vddio3"),
+#ifndef CFG_STM32MP21
 	[IOD_VDDIO4] = DEFINE_REGUL(IOD_VDDIO4, "vddio4", "vddio4"),
+#endif
 	[IOD_VDDIO]  = DEFINE_REGUL(IOD_VDDIO, "vddio", "vdd"),
+#ifndef CFG_STM32MP21
 	[REGU_UCPD]  = DEFINE_FIXED(REGU_UCPD, "vdd33ucpd", "vdd33ucpd"),
+#endif
 	[REGU_A]     = DEFINE_FIXED(REGU_A, "vdda18adc", "vdda18adc"),
+#ifndef CFG_STM32MP21
 	[REGU_GPU]   = DEFINE_REGUL(REGU_GPU, "vddgpu", "vddgpu"),
+#endif
 };
 DECLARE_KEEP_PAGER(pwr_regu_desc);
 
-- 
2.47.2

